### PR TITLE
[Feature] Add `OpenSpielWrapper` and `OpenSpielEnv`

### PR DESCRIPTION
**Pull request by kurtamohler**
## Description

Adds environment wrapper classes for OpenSpiel.

`OpenSpielWrapper.reset` supports resetting to a specified state.

## Motivation and Context

Part of #2133

- [x] Someone has raised an issue to propose this change ([required](https://github.com/pytorch/rl/issues) for new features and bug fixes)

## Types of changes

What types of changes does your code introduce? Remove all that do not apply:

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds core functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [x] Documentation (update in the documentation)
- [ ] Example (update in the folder of examples)

## Checklist

Go over all the following points, and put an `x` in all the boxes that apply.
If you are unsure about any of these, don't hesitate to ask. We are here to help!

- [x] I have read the [CONTRIBUTION](https://github.com/pytorch/rl/blob/main/CONTRIBUTING.md) guide (**required**)
- [x] My change requires a change to the documentation.
- [x] I have updated the tests accordingly (*required for a bug fix or a new feature*).
- [x] I have updated the documentation accordingly.

### CODE REVIEW

**Comment by kurtamohler** [comment]
A few notes:

`categorical_action_encoding=False` is not supported yet.

Also, some of the games in OpenSpiel do not work properly in `OpenSpielWrapper` because the action spec assumes a discrete space of `pyspiel.Game.num_distinct_actions()` (see [OpenSpiel API reference](https://github.com/google-deepmind/open_spiel/blob/82c742c6403aefeebb104736387c19d709fa7255/docs/api_reference.md)). However, for some of the games, `pyspiel.State.legal_actions()` can return more actions than `pyspiel.Game.num_distinct_actions()`. I suppose to support those games we need to allow the action spec's size to change?

File: torchrl/envs/libs/openspiel.py [hunk 1]
```diff
@@ -0,0 +1,358 @@
+import importlib.util
```
Comment (vmoens) [inline torchrl/envs/libs/openspiel.py:1]: missing headers

File: torchrl/envs/libs/openspiel.py [hunk 2]
```diff
@@ -0,0 +1,358 @@
...
+from typing import Dict, Optional
```
Comment (vmoens) [inline torchrl/envs/libs/openspiel.py:2]: I try to upgrade to the new `foo: Smth | SmthElse` type annotation, can we remove the `Optional` and use that instead (make sure you have a `from __future__ import annotations` for older python versions)

File: torchrl/envs/libs/openspiel.py [hunk 3]
```diff
@@ -0,0 +1,358 @@
...
+    Args:
+        env (pyspiel.State): the game to wrap.
+        categorical_action_encoding (bool, optional): if ``True``, categorical
+            specs will be converted to the TorchRL equivalent (:class:`torchrl.data.DiscreteTensorSpec`),
+            otherwise a one-hot encoding will be used (:class:`torchrl.data.OneHotTensorSpec`).
+            Defaults to ``False``.
+
+    Keyword Args:
+        device (torch.device, optional): if provided, the device on which the data
+            is to be cast. Defaults to ``torch.device("cpu")``.
```
Comment (vmoens) [inline torchrl/envs/libs/openspiel.py:46]: device should be `None` by default

File: torchrl/envs/libs/openspiel.py [hunk 4]
```diff
@@ -0,0 +1,358 @@
...
+                    observation_shape
+                )
+                for idx in range(self._env.num_players())
+            ]
+        )
+        reward = self._to_tensor(self._env.returns())
+        done = torch.tensor(
+            self._env.is_terminal(), device=self.device, dtype=torch.bool
+        )
+        state = self._env.serialize()
```
Comment (vmoens) [inline torchrl/envs/libs/openspiel.py:206]: let's put a flag around that to make the state optional. Maybe "stateless=True"?

File: torchrl/envs/libs/openspiel.py [hunk 5]
```diff
@@ -0,0 +1,358 @@
...
+
+    def _update_action_mask(self):
+        action_mask = torch.zeros(
+            self._env.num_distinct_actions(), device=self.device, dtype=torch.bool
+        )
+        action_mask[self._env.legal_actions()] = True
+        self.action_spec.update_mask(action_mask)
+
+    def _make_td_out(self, exclude_reward=False):
+        observation_shape = self._env.get_game().observation_tensor_shape()
```
Comment (vmoens) [inline torchrl/envs/libs/openspiel.py:193]: How expensive is `get_game()`? Can we cache it?

File: torchrl/envs/libs/openspiel.py [hunk 6]
```diff
@@ -0,0 +1,358 @@
...
+            state=NonTensorSpec([]),
+            shape=self.batch_size,
+        )
+
+    def _set_seed(self, seed):
+        if seed is not None:
+            raise NotImplementedError("This environment has no seed.")
+
+    def current_player(self):
+        return self._env.current_player()
```
Comment (vmoens) [inline torchrl/envs/libs/openspiel.py:183]: @matteobettini if we have multiple players, do we have to consider that as a MARL problem?
What's the proper way to represent the env here?

I see various options:
- what we are currently doing, tracking the player in the TD
- Having a nested TD with `"player0"` and `"player1"` (should that be dynamic)
- i also thought: can we have 2 envs that play against each other? I guess for MCTS that could be nice to keep things separated

File: torchrl/envs/libs/openspiel.py [hunk 7]
```diff
@@ -0,0 +1,358 @@
...
+                "terminated": done.clone(),
+                "state": state,
+                "current_player": current_player,
+            },
+            batch_size=self.batch_size,
+            device=self.device,
+        )
+
+        if not exclude_reward:
+            tensordict_out["reward"] = reward
```
Comment (vmoens) [inline torchrl/envs/libs/openspiel.py:224]: setting it after the fact is more expensive than putting it in the dict or not, maybe we could do 
```python
source = {
                "observation": observation,
                "done": done,
                "terminated": done.clone(),
                "state": state,
                "current_player": current_player,
            }
if not exclude_reward:
    source["reward"] = reward
tensordict_out = TensorDict(source, ...)
```

File: torchrl/envs/libs/openspiel.py [hunk 8]
```diff
@@ -0,0 +1,358 @@
...
+        game = self._env.get_game()
+
+        if tensordict is not None and "state" in tensordict:
+            new_env = game.deserialize_state(tensordict["state"])
+        else:
+            new_env = game.new_initial_state()
+
+        self._env = new_env
+        self._update_action_mask()
+        return self._make_td_out(exclude_reward=True)
```
Comment (vmoens) [inline torchrl/envs/libs/openspiel.py:254]: interesting, so openspiel has a reward at reset time?

**Review by vmoens — COMMENTED** [review]
Great work thanks for that, I left a couple of comments already.
Is the TODO blocking on this?

File: torchrl/envs/libs/openspiel.py (hunk 6 above)
Comment (matteobettini) [reply torchrl/envs/libs/openspiel.py:183]: In pettingzoo we have some boatd games already. I guess the choice here also depends on if the game is turn-based or not (players can  act in parallel)

the way we did it in pettingzoo is to keep the players in 2 nested td groups (your option 2) and have a mask for the non-playing one.

Comment (vmoens) [reply torchrl/envs/libs/openspiel.py:183]: the goal is to use it for MCTS so I'd say turn based.
Self-play sort of thing

File: torchrl/envs/libs/openspiel.py (hunk 5 above)
Comment (kurtamohler) [reply torchrl/envs/libs/openspiel.py:193]: It isn't very expensive, but I tried caching it with `functools.lru_cache`, and it does speed it up by about 6x, so I'll update it.

```python
import timeit
iters = 1_000_000
res0 = timeit.timeit(
    setup='import torchrl; env = torchrl.envs.OpenSpielEnv("chess")',
    stmt='env._env.get_game()',
    number=iters)
res1 = timeit.timeit(
    setup='import torchrl; env = torchrl.envs.OpenSpielEnv("chess")',
    stmt='env._get_game_cached()',
    number=iters)
print(f'nocache: {res0 / iters} s')
print(f'cache: {res1 / iters} s')
print(f'speedup: {res0 / res1} x')
```

```
nocache: 4.2987480599913395e-07 s
cache: 6.127979099983349e-08 s
speedup: 7.014952221366127x
```

File: torchrl/envs/libs/openspiel.py (hunk 8 above)
Comment (kurtamohler) [reply torchrl/envs/libs/openspiel.py:254]: Yes, but it's just set to zero

File: torchrl/envs/libs/openspiel.py (hunk 4 above)
Comment (kurtamohler) [reply torchrl/envs/libs/openspiel.py:206]: What if we call the flag `output_state`? I think that name makes it more obvious what exactly it does--it adds the state to the output of `reset` and `step`

File: torchrl/envs/libs/openspiel.py (hunk 5 above)
Comment (kurtamohler) [reply torchrl/envs/libs/openspiel.py:193]: Well, flake8 fails, pointing out that `lru_cache` on an instance method can cause memory leaks: https://docs.astral.sh/ruff/rules/cached-instance-method/

So I decided to just cache it manually using an initially-None data member. Speedup decreased to 5x, but that still makes it worth doing.

File: torchrl/envs/libs/openspiel.py (hunk 6 above)
Comment (kurtamohler) [reply torchrl/envs/libs/openspiel.py:247]: I will push an update to put the players in their own nested td groups with a mask, like in PettingZooEnv

**Comment by kurtamohler** [comment]
At the moment, this only supports games where all actions are taken by the players, like in chess or tic-tac-toe. But I've realized that OpenSpiel also has a concept of chance nodes, where a random non-player action is taken. For instance, in Kuhn poker, the initial dealing of the cards is a chance node. In liar's dice, the outcome of rolling dice is a chance node. OpenSpiel has some methods to obtain all the possible chance actions and the associated probability distribution ([shown in this example](https://openspiel.readthedocs.io/en/latest/concepts.html#playing-a-trajectory)).

For now, I'll raise an error if a loaded game contains chance nodes and leave it as a TODO for me to add support for it in a follow-up PR--unless you would prefer for me to add it in this PR.

File: torchrl/envs/libs/openspiel.py (hunk 4 above)
Comment (kurtamohler) [reply torchrl/envs/libs/openspiel.py]: I renamed to `return_state`, because that's what `PettingZooWrapper` calls it

**Comment by vmoens** [comment]
> I suppose to support those games we need to allow the action spec's size to change at each step?

Yes I think having a dynamic space would be the way to go. See #2143

**Comment by vmoens** [comment]
> OpenSpiel has some methods to obtain all the possible chance actions and the associated probability distribution (shown in this example).

That's an amazing feature. Happy to integrate it separately!

**Comment by kurtamohler** [comment]
> > I suppose to support those games we need to allow the action spec's size to change at each step?
> 
> Yes I think having a dynamic space would be the way to go. See #2143

Actually, I'm not so sure that we would need dynamic action specs after all. It is true that `pyspiel.State.legal_actions()` can return a different number of actions than `pyspiel.Game.num_distinct_actions()`, but I'm pretty sure (not 100% sure) that only happens when it's a chance node, in which case the length of `legal_actions()` is `pyspiel.Game.max_chance_outcomes()` instead. Once I add support for chance nodes, it will have its own action spec separate from the players' action specs, and I think all action specs will maintain the same shape throughout the game.

**Comment by kurtamohler** [comment]
I think I've addressed everything that needed to be fixed so far. Let me know if there is anything else

**Review by vmoens — APPROVED** [review]
Amazing thanks a mil for this!
